### PR TITLE
KNN: Fix optional parameters and ordering

### PR DESCRIPTION
**Pull request by orhankislal**
Additional Author : Frank McQuillan <fmcquillan@pivotal.io>

JIRA: MADLIB-1129

Add the necessary interfaces for optional parameters.
Add ordering for the nearest neighbors.
Change the default for output_neighbors to True.
Update the documentation to reflect the changes.

### CODE REVIEW

**Comment by fmcquillan99** [comment]
LGTM based on some testing and docs review.
I would ask other community folks to pls review code in more detail however.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/260/

**Comment by orhankislal** [comment]
jenkins ok to test

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/261/

File: src/ports/postgres/modules/knn/knn.py_in [hunk 1]
```diff
@@ -135,13 +135,17 @@ def knn(schema_madlib, point_source, point_column_name, point_id, label_column_n
         test_id_temp = unique_string(desp='test_id_temp')
 
         if output_neighbors is None or '':
-            output_neighbors = False
+            output_neighbors = True
```
[the remark this answers is not in the document]
Comment (iyerr3) [reply src/ports/postgres/modules/knn/knn.py_in]: On second read of the code: output_neighbors is already a boolean. Is there a reason to expect it to be `''`?

File: src/ports/postgres/modules/knn/knn.py_in [hunk 2]
```diff
@@ -135,13 +135,17 @@ def knn(schema_madlib, point_source, point_column_name, point_id, label_column_n
...
 
         interim_table = unique_string(desp='interim_table')
 
         if label_column_name is None or label_column_name == '':
```
Comment (iyerr3) [inline src/ports/postgres/modules/knn/knn.py_in]: `if not label_column_name`

File: src/ports/postgres/modules/knn/knn.py_in [hunk 3]
```diff
@@ -135,13 +135,17 @@ def knn(schema_madlib, point_source, point_column_name, point_id, label_column_n
...
-            plpy.execute(
-                """
+
+            if output_neighbors is False:
+                plpy.error("kNN error: Either label_column_name or "
+                       "output_neighbors has to be non-NULL.")
+
+            sql1 = """
```
Comment (iyerr3) [inline src/ports/postgres/modules/knn/knn.py_in]: if `sql1` and `sql2` (below) are not reused, then I would prefer to just directly call the `plpy.execute` with the string.

File: src/ports/postgres/modules/knn/knn.py_in [hunk 4]
```diff
@@ -215,7 +222,8 @@ def knn(schema_madlib, point_source, point_column_name, point_id, label_column_n
         knn_pred_class = schema_madlib + \
             '.mode(' + label_col_temp + ') AS prediction'
         knn_pred_reg = 'avg(' + label_col_temp + ') AS prediction'
-        knn_neighbours = ', array_agg(knn_temp.train_id) AS k_nearest_neighbours '
+        knn_neighbours = ', array_agg(knn_temp.train_id ORDER BY knn_temp.dist ASC)'\
```
Comment (iyerr3) [inline src/ports/postgres/modules/knn/knn.py_in]: Preferred to break long lines by wrapping expressions in parentheses over using a backslash. ([PEP8 link](https://www.python.org/dev/peps/pep-0008/#maximum-line-length))

**Comment by iyerr3** [comment]
jenkins ok to retest

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/263/

**Comment by orhankislal** [comment]
Thanks for the comments @jingyimei and @iyerr3. I have made some significant changes to the code with the latest commit. I would greatly appreciate your comments on them.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/264/

**Pull request closed without merging**